rebalance/rebalance.py

```diff
@@ -212,16 +215,260 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi=None,
+    their_liquidity = channel["msatoshi_total"] - channel["msatoshi_to_us"]
+    return max(0, their_liquidity - min_liquidity)
+
+
+def get_our_channels(plugin: Plugin):
+    channels = []
+    for peer in plugin.rpc.listpeers()["peers"]:
+        for ch in peer["channels"]:
+            if ch["state"] == "CHANNELD_NORMAL" and not ch["private"]:
+                channels.append(ch)
+    return channels
+
+
+def check_liquidity_threshold(channels: list, threshold: int):
+    # check if overall rebalances can be successful with this threshold
+    ours = sum(ch["msatoshi_to_us"] for ch in channels)
+    total = sum(ch["msatoshi_total"] for ch in channels)
+    required = 0
+    for ch in channels:
+        required += int(min(threshold, ch["msatoshi_total"] / 2))
+    return required < ours and required < total - ours
+
+
+def binary_search(channels: list, low: int, high: int):
+    if high - low < 1000:
+        return low
+    next_step = int((low + high) / 2)
+    if check_liquidity_threshold(channels, next_step):
+        return binary_search(channels, next_step, high)
+    else:
+        return binary_search(channels, low, next_step)
+
+
+def get_enough_liquidity_threshold(channels: list):
+    biggest_channel = max(channels, key=lambda ch: ch["msatoshi_total"])
+    max_threshold = binary_search(channels, 0, int(biggest_channel["msatoshi_total"] / 2))
+    return int(max_threshold / 2)
+
+
+def get_ideal_ratio(channels: list, enough_liquidity: int):
+    # ideal liquidity ratio for big channels:
+    # small channels should have a 50/50 liquidity ratio to be usable
+    # and big channels can store the remaining liquidity above the threshold
+    ours = sum(ch["msatoshi_to_us"] for ch in channels)
+    total = sum(ch["msatoshi_total"] for ch in channels)
+    chs = list(channels)
+    while True:
+        ratio = ours / total
+        smallest_channel = min(chs, key=lambda ch: ch["msatoshi_total"])
+        if smallest_channel["msatoshi_total"] * min(ratio, 1 - ratio) > enough_liquidity:
+            break
+        min_liquidity = min(smallest_channel["msatoshi_total"] / 2, enough_liquidity)
+        diff = smallest_channel["msatoshi_total"] * ratio
+        diff = max(diff, min_liquidity)
+        diff = min(diff, smallest_channel["msatoshi_total"] - min_liquidity)
+        ours -= diff
+        total -= smallest_channel["msatoshi_total"]
+        chs.remove(smallest_channel)
+    return ratio
+
+
+def feeadjust_would_be_nice(plugin: Plugin):
+    try:
+        msg = plugin.rpc.feeadjust()
+        plugin.log(f"Feeadjust succeeded: {msg}")
+    except Exception:
+        plugin.log("The feeadjuster plugin would be useful here")
+
+
+def get_max_amount(i: int, min_amount: int, enough_liquidity: int):
+    return max(min_amount, int(enough_liquidity / (4**(i + 1))))
+
+
+def get_max_fee(plugin: Plugin, msat: int):
+    # TODO: sanity check
+    return int((plugin.basefee + plugin.ppmfee * msat / 1000000) * plugin.feeratio)
```

Comment:
If possible, use `Millisatoshi` and get rid of the `int()` casts below
```suggestion
def get_max_fee(plugin: Plugin, msat: Millisatoshi):
    # TODO: sanity check
    return (plugin.basefee + msat * plugin.ppmfee / 1000000) * plugin.feeratio
```